proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +156,481 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+
+RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      MatrixC = {RWBuffer, /*offset=*/64, /*stride=*/0};
+```
+
+Template parameters:
+
+- `DT` - the type used to store elements of the matrix in memory
+- `M` - the 'M' dimension of the matrix
+- `K` - the 'K" dimension of the matrix
+- `ML` - the layout of the matrix in memory
+- `TRANPOSE` - whether or not this matrix should be transposed before any operations
+
+Members:
+
+- `Buffer` - the buffer that the matrix is stored in
+  - For `MatrixRef` this is a `ByteAddressBuffer`
+  - For `RWMatrixRef` this is a `RWByteAddresssBuffer`
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer where
+  the matrix is located.
+- `Stride` - the stride, in bytes, between rows/columns of the matrix. This
+  value is ignored if the matrix layout is `MATRIX_LAYOUT_MUL_OPTIMAL` or
+  `MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL`.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BUFFER, DataType DT, uint M, uint K, MatrixLayout ML,
+          bool TRANSPOSE>
+struct MatrixRefImpl {
+  BUFFER Buffer;
+  uint StartOffset;
+  uint Stride;
+};
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool TRANSPOSE = false>
+using MatrixRef = MatrixRefImpl<ByteAddressBuffer, DT, M, K, ML, TRANSPOSE>;
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool TRANSPOSE = false>
+using RWMatrixRef = MatrixRefImpl<RWByteAddressBuffer, DT, M, K, ML, TRANSPOSE>;
+
+} // namespace linalg
+} // namespace dx
+```
+
+### struct VectorRef
+
+`VectorRef` and `RWVectorRef` specify a reference to a vector in memory along
+with the format of each element.
+
+As with `MatrixRef`, two versions are provided - one for vectors stored in
+`ByteAddressBuffer` and another for `RWByteAddressBuffer`. A base class,
+`VectorRefImpl`, covers both of them.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+using namesace dx::linalg;
+
+VectorRef<DATA_TYPE_FLOAT16> VectorA = {ROBuffer, /*offset=*/128};
+VectorRef<DATA_TYPE_FLOAT32> VectorB = {ROBuffer, /*offset=*/128};
+RWVectorRef<DATA_TYPE_SINT16> VectorC = {RWBuffer, /*offset=*/64};
+```
+
+Template parameter:
+
+- `DT` - the data type of each element stored in the buffer
+
+Members:
+
+- `Buffer` - the buffer that the vector is stored in
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer to
+  where the vector is located.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BUFFER, DataType DT>
+struct VectorRefImpl {
+  BUFFER Buffer;
+  uint StartOffset;
+};
+
+template <DataType DT>
+using VectorRef = VectorRefImpl<ByteAddressBuffer, DT>;
+
+template <DataType DT>
+using RWVectorRef = VectorRefImpl<RWByteAddressBuffer, DT>;
+
+} // namespace linalg
+} // namespace dx
+
+```
+
+### struct Vector
+
+> NOTE: it's possible that one resolution of [441] may remove the need for this
```

Comment:
```suggestion
> NOTE: it's possible that one resolution of
> [441](https://github.com/microsoft/hlsl-specs/issues/441) may remove the need for this
```

Reply:
I see a few suggestions for inlining the links - but this document is using reference-style links.  Is the overall suggestion here to move away from reference-style links?

Reply:
Oh, this is fine. I always forget about reference linking and that you don't actually need to list a reference text that is different from the displayed text.